### PR TITLE
feat: use OIDC server with fine-grained authorization

### PR DESCRIPTION
**Pull request by vincentchalamon**
### Description

tl;dr: this PR aims to deport the authorization out of Symfony to the OIDC server.

Now that we have an OIDC server, and in order to respect the decentralization state of the art, the Symfony application should not handle the user roles and permissions anymore, as it's supposed to be the responsability of the OIDC server. The Symfony application is a resource server which communicates with the OIDC server to decentralize the user roles and permissions.

To implement this enhancement, the Symfony application must evolve to contact the OIDC server in order to check the user roles and permissions:
- _user roles_ are global roles applied on a user, for instance: `user` or `admin`
- _user permissions_ are user permissions on a resource, for instance: `owner of *this* review`

### TODO

- [x] create voter to check for user roles from token (`UserRoleVoter`)
- [x] create voter to check for user roles from OIDC server (`UserTokenIntrospectRoleVoter`)
- [x] create voter to check for user permissions from OIDC server (`UserTokenPermissionVoter`)
- [x] create Review resources on OIDC server for remote permissions checking

Improvements (could be backported to Symfony):

- [x] Use OIDC Discovery with local cache (OIDC configuration + JWKSet)

### Links

https://www.keycloak.org/docs/latest/authorization_services/index.html

### CODE REVIEW

File: api/src/Entity/Bookmark.php [hunk 1]
```diff
@@ -36,7 +36,7 @@
     operations: [
         new GetCollection(),
         new Delete(
-            security: 'is_granted("ROLE_USER") and object.user === user'
+            security: 'object.user == user'
```
Comment (soyuka) [inline api/src/Entity/Bookmark.php:39]: why don't we need `is_granted("ROLE_USER")` anymore?

File: api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    public function getUserBadgeFrom(string $accessToken): UserBadge
+    {
+        if (!class_exists(JWSVerifier::class) || !class_exists(Checker\HeaderCheckerManager::class)) {
+            throw new \LogicException('You cannot use the "oidc_discovery" token handler since "web-token/jwt-signature" and "web-token/jwt-checker" are not installed. Try running "composer require web-token/jwt-signature web-token/jwt-checker".');
+        }
+
+        if (!class_exists(HttpClient::class)) {
+            throw new \LogicException('You cannot use the "oidc_discovery" token handler since "symfony/http-client" is not installed. Try running "composer require symfony/http-client".');
+        }
```
Comment (soyuka) [inline api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php:52]: this is not a library maybe we can get rid of these 2 ifs?

File: api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        $oidcConfiguration = json_decode($this->cache->get('oidc.configuration', function (ItemInterface $item): string {
+            $item->expiresAfter($this->ttl);
+            $response = HttpClient::create()->request('GET', rtrim($this->issuer, '/') . '/.well-known/openid-configuration');
```
Comment (soyuka) [inline api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php:56]: shouldn't you inject an HttpClient instead?

File: api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+            return $response->getContent();
+        }), true, 512, \JSON_THROW_ON_ERROR);
+
+        $keyset = JWKSet::createFromJson(
+            $this->cache->get('oidc.jwkSet', function (ItemInterface $item) use ($oidcConfiguration): string {
+                $item->expiresAfter($this->ttl);
+                $client = HttpClient::create();
```
Comment (soyuka) [inline api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php:64]: same why not injecting an http client?

File: api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                $response = $client->request('GET', $oidcConfiguration['jwks_uri']);
+                // we only need signature key
+                $keys = array_filter($response->toArray()['keys'], static fn (array $key) => 'sig' === $key['use']);
+
+                return json_encode(['keys' => $keys]);
+            })
+        );
```
Comment (soyuka) [inline api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php:71]: It doesn't look easy but we should probably try/catch a bunch of errors no? I'm thinking about the `toArray()`(what happens if `jwks_uri` is wrong or unreacheable? Above, the `JSON_THROW_ON_ERROR` isn't catched. (I started reviewing this so maybe that the tokenhandler exceptions are handled in another file, please ignore if that's the case)

File: api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+        try {
+            // Decode the token
+            $signature = 0;
```
Comment (soyuka) [inline api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php:75]: why is the signature `0` ?

File: api/src/Security/Voter/OIDCPermissionVoter.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                    'audience' => $this->oidcClientId,
+                    'response_mode' => 'decision',
+                    'permission_resource_format' => 'uri',
+                    'permission_resource_matching_uri' => true,
+                    'permission' => sprintf('%s', $subject),
+                ],
+            ]);
+
+            return $response->toArray()['result'] ?? false;
+        } catch (ExceptionInterface) {
```
Comment (soyuka) [inline api/src/Security/Voter/OIDCPermissionVoter.php:70]: shouldn't you log the exception ?

File: api/src/Security/Voter/OIDCRoleVoter.php [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                    'client_secret' => $this->oidcClientSecret,
+                    'token' => $accessToken,
+                ],
+            ]);
+
+            $roles = array_map(static fn (string $role): string => strtolower($role), $response->toArray()['realm_access']['roles'] ?? []);
+        } catch (HttpExceptionInterface) {
+            // OIDC server said no!
+            return false;
+        } catch (ExceptionInterface) {
```
Comment (soyuka) [inline api/src/Security/Voter/OIDCRoleVoter.php:69]: shouldn't you log exceptions?

File: api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php (hunk 6 above)
Comment (vincentchalamon) [reply api/src/Security/Http/AccessToken/Oidc/OidcDiscoveryTokenHandler.php:75]: Required by web-token/jwt-library, `$signature` is passed by reference and required... But it can be `null` as default value.
